src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -229,152 +298,651 @@ ValidatorList::CacheValidatorFile(
-            msg.set_manifest(manifest);
-            msg.set_blob(blob);
-            msg.set_signature(signature);
-            msg.set_version(version);
-
-            auto const& publisherKey = *result.publisherKey;
-            auto const sequence = *result.sequence;
-
-            // Can't use overlay.foreach here because we need to modify
-            // the peer, and foreach provides a const&
-            auto message =
-                std::make_shared<Message>(msg, protocol::mtVALIDATORLIST);
-            for (auto& peer : overlay.getActivePeers())
+            (void)pubKey;
+            if (listCollection.status != PublisherStatus::available)
             {
-                if (toSkip->count(peer->id()) == 0 &&
-                    peer->supportsFeature(
-                        ProtocolFeature::ValidatorListPropagation) &&
-                    peer->publisherListSequence(publisherKey) < sequence)
-                {
-                    peer->send(message);
-
-                    JLOG(j_.debug())
-                        << "Sent validator list for " << strHex(publisherKey)
-                        << " with sequence " << sequence << " to "
-                        << peer->getRemoteAddress().to_string() << " ("
-                        << peer->id() << ")";
-                    // Don't send it next time.
-                    hashRouter.addSuppressionPeer(hash, peer->id());
-                    peer->setPublisherListSequence(publisherKey, sequence);
-                }
+                good = false;
+                break;
             }
         }
+        if (good)
+        {
+            networkOPs.clearUNLBlocked();
+        }
+    }
+    bool broadcast = disposition <= ListDisposition::known_sequence;
+
+    if (broadcast)
+    {
+        auto const& pubCollection = publisherLists_[*result.publisherKey];
+        assert(
+            result.status <= PublisherStatus::expired && result.publisherKey &&
+            pubCollection.maxSequence);
+        broadcastBlobs(
+            *result.publisherKey,
+            pubCollection,
+            *pubCollection.maxSequence,
+            hash,
+            overlay,
+            hashRouter,
+            j_);
     }
 
     return result;
 }
 
 ValidatorList::PublisherListStats
-ValidatorList::applyList(
+ValidatorList::applyLists(
     std::string const& manifest,
-    std::string const& blob,
-    std::string const& signature,
     std::uint32_t version,
+    std::vector<ValidatorBlobInfo> const& blobs,
     std::string siteUri,
-    boost::optional<uint256> const& hash)
+    boost::optional<uint256> const& hash /* = {} */)
 {
-    using namespace std::string_literals;
-
-    if (version != requiredListVersion)
+    if (std::count(
+            std::begin(supportedListVersions),
+            std::end(supportedListVersions),
+            version) != 1)
         return PublisherListStats{ListDisposition::unsupported_version};
 
     std::unique_lock<std::shared_timed_mutex> lock{mutex_};
```

Comment:
Not in this change set, but it seems like this should be a `lock_guard`, not a `unique_lock`.

Reply:
That's gonna be fun to merge with develop, but, yeah, it makes sense.

Reply:
Fixed after rebasing